src/pubsub.js

```diff
@@ -0,0 +1,573 @@
+/* eslint-env mocha */
+/* eslint max-nested-callbacks: ['error', 8] */
+'use strict'
+
+const expect = require('chai').expect
+const series = require('async/series')
+const waterfall = require('async/waterfall')
+const parallel = require('async/parallel')
+const whilst = require('async/whilst')
+const each = require('async/each')
+
+function waitForPeers (ipfs, topic, peersToWait, callback) {
+  const i = setInterval(() => {
+    ipfs.pubsub.peers(topic, (err, peers) => {
+      if (err) {
+        return callback(err)
+      }
+
+      const missingPeers = peersToWait
+            .map((e) => peers.includes(e))
+            .filter((e) => !e)
+
+      if (missingPeers.length === 0) {
+        clearInterval(i)
+        callback()
+      }
+    })
+  }, 500)
+}
+
+function spawnWithId (factory, callback) {
+  waterfall([
+    (cb) => factory.spawnNode(cb),
+    (node, cb) => node.id((err, res) => {
+      if (err) {
+        return cb(err)
+      }
+      node.peerId = res
+      cb(null, node)
+    })
+  ], callback)
+}
+
+function makeCheck (n, done) {
+  let i = 0
+  return (err) => {
+    if (err) {
+      return done(err)
+    }
+
+    if (++i === n) {
+      done()
+    }
+  }
+}
+
+module.exports = (common) => {
+  describe('.pubsub', () => {
+    const topic = 'pubsub-tests'
+
+    describe('callback API', () => {
+      let ipfs1
+      let ipfs2
+      let ipfs3
+
+      before((done) => {
+        common.setup((err, factory) => {
+          if (err) {
+            return done(err)
+          }
+
+          series([
+            (cb) => spawnWithId(factory, cb),
+            (cb) => spawnWithId(factory, cb),
+            (cb) => spawnWithId(factory, cb)
+          ], (err, nodes) => {
+            if (err) {
+              return done(err)
+            }
+
+            ipfs1 = nodes[0]
+            ipfs2 = nodes[1]
+            ipfs3 = nodes[2]
+            done()
+          })
+        })
+      })
+
+      after((done) => {
+        common.teardown(done)
+      })
+
+      describe('single node', () => {
+        describe('.publish', () => {
+          it('message from string', (done) => {
+            ipfs1.pubsub.publish(topic, 'hello friend', done)
+          })
```

Comment:
I'm worried with accepting plain strings as argument. The issue that we had last week Monday was because the initial pubsub code @haadcode wrote would always convert any value to a Buffer on the publish implementation, but then, because the tests were written like `expect(dataReceived).to.equal('some string')`, the implementation had a `data.toString()` internally as a quick patch, and when I changed that to make it an operation in the outside (like we have here now, see below, preserving the buffer type), it made orbit stop working because there was an assumption that 'data' was a string, since orbit publishes strings (or was publishing).

Strings and Buffers are one of banes of JavaScript, it would be better to force it to be just Buffer and no other type (creating a test here to make sure it fails with anything non buffer), to avoid future problems.